serve/methods/gas.go

```diff
@@ -9,82 +9,90 @@ import (
 )
 
 type gasFeeTotalInfo struct {
-	Low         int64
-	High        int64
-	TotalAmount int64
+	Low         float64
+	High        float64
+	TotalAmount float64
 	TotalCount  int64
 }
 
-// GetGasPricesByBlock calculates the gas price statistics (low, high, average)
-// for a single block.
-func GetGasPricesByBlock(block *types.Block) ([]*GasPrice, error) {
-	blocks := []*types.Block{block}
+// GetGasPricesByTxResult calculates the gas price statistics (low, high, average)
+// for a single txResult.
+func GetGasPricesByTxResult(txResult *types.TxResult) ([]*GasPrice, error) {
+	txResults := []*types.TxResult{txResult}
 
-	return GetGasPricesByBlocks(blocks)
+	return GetGasPricesByTxResults(txResults)
 }
 
-// GetGasPricesByBlocks calculates the gas price statistics (low, high, average)
-// for multiple blocks.
-func GetGasPricesByBlocks(blocks []*types.Block) ([]*GasPrice, error) {
+// GetGasPricesByTxResults calculates the gas price statistics (low, high, average)
+// for multiple txResults.
+func GetGasPricesByTxResults(txResults []*types.TxResult) ([]*GasPrice, error) {
 	gasFeeInfoMap := make(map[string]*gasFeeTotalInfo)
 
-	for _, block := range blocks {
-		blockGasFeeInfo := calculateGasFeePerBlock(block)
+	txResultGasFeeInfo := calculateGasFeePerTxResults(txResults)
 
-		for denom, gasFeeInfo := range blockGasFeeInfo {
-			_, exists := gasFeeInfoMap[denom]
-			if !exists {
-				gasFeeInfoMap[denom] = &gasFeeTotalInfo{}
-			}
+	for denom, gasFeeInfo := range txResultGasFeeInfo {
+		_, exists := gasFeeInfoMap[denom]
+		if !exists {
+			gasFeeInfoMap[denom] = &gasFeeTotalInfo{}
+		}
 
-			err := modifyAggregatedInfo(gasFeeInfoMap[denom], gasFeeInfo)
-			if err != nil {
-				return nil, err
-			}
+		err := modifyAggregatedInfo(gasFeeInfoMap[denom], gasFeeInfo)
+		if err != nil {
+			return nil, err
 		}
 	}
 
 	return calculateGasPrices(gasFeeInfoMap), nil
 }
 
-// calculateGasFeePerBlock processes all transactions in a single block to compute
+// calculateGasFeePerTxResult processes all transactions in a single txResult to compute
 // gas fee statistics (low, high, total amount, total count) for each gas fee denomination.
-func calculateGasFeePerBlock(block *types.Block) map[string]*gasFeeTotalInfo {
+func calculateGasFeePerTxResults(txResults []*types.TxResult) map[string]*gasFeeTotalInfo {
 	gasFeeInfo := make(map[string]*gasFeeTotalInfo)
 
-	for _, t := range block.Txs {
+	for _, t := range txResults {
+		if t.Response.IsErr() ||
+			t.Response.GasUsed == 0 ||
+			t.Height <= 0 {
+			continue
+		}
+
 		var stdTx std.Tx
-		if err := amino.Unmarshal(t, &stdTx); err != nil {
+		if err := amino.Unmarshal(t.Tx, &stdTx); err != nil {
 			continue
 		}
 
 		denom := stdTx.Fee.GasFee.Denom
-		amount := stdTx.Fee.GasFee.Amount
+		gasFeeAmount := stdTx.Fee.GasFee.Amount
+		GasUsedAmount := t.Response.GasUsed
+
+		// Calculate the gas price (gasFee / gasUsed)
+		gasPrice := float64(gasFeeAmount) / float64(GasUsedAmount)
```

Comment:
The gas used has no influence on the gas price, actually. The chain takes everything the user specified in `gasFee` as part of the tx, even though the tx itself took less gas than what the user specified in `GasUsedAmount`

See:
https://github.com/gnolang/gno/issues/3805